NHapi20/NHapi.Base/Model/AbstractSegment.cs

```diff
@@ -1,423 +1,487 @@
+		 {
+			throw new HL7Exception(
+				"Can't retrieve cardinality of field " + number + " from segment " + GetType().FullName + " - there are only " +
+				_items[number - 1].Fields.Count + " fields.", HL7Exception.APPLICATION_INTERNAL_ERROR);
+		 }
+
+		 int reps = 0;
+		 try
+		 {
+			reps = _items[number - 1].MaxRepetitions; //fields #d from 1 to user
+		 }
+		 catch (Exception e)
+		 {
+			throw new HL7Exception("Can't retrieve max repetitions of field " + number + ": " + e.Message,
+				HL7Exception.APPLICATION_INTERNAL_ERROR);
+		 }
+
+		 return reps;
+	  }
+
+	  /// <summary> Adds a field to the segment.  The field is initially empty (zero repetitions).
+	  /// The field number is sequential depending on previous add() calls.  Implementing
+	  /// classes should use the add() method in their constructor in order to define fields
+	  /// in their segment.
+	  /// </summary>
+	  /// <param name="c">the class of the data for this field - this should inherit from Type
+	  /// </param>
+	  /// <param name="required">whether a value for this field is required in order for the segment
+	  /// to be valid
+	  /// </param>
+	  /// <param name="maxReps">the maximum number of repetitions - 0 implies that there is no limit
+	  /// </param>
+	  /// <param name="length">the maximum length of each repetition of the field (in characters)
+	  /// </param>
+	  /// <param name="constructorArgs">an array of objects that will be used as constructor arguments
+	  /// if new instances of this class are created (use null for zero-arg constructor)
+	  /// </param>
+	  /// <throws>  HL7Exception if the given class does not inherit from Type or if it can  </throws>
+	  /// <summary>    not be instantiated.
+	  /// </summary>
+	  protected internal virtual void add(Type c, bool required, int maxReps, int length, Object[] constructorArgs)
+	  {
+		 add(c, required, maxReps, length, constructorArgs, null);
+	  }
+
+	  /// <summary>
+	  /// Add a segment
+	  /// </summary>
+	  /// <param name="c">The type of segment</param>
+	  /// <param name="required"></param>
+	  /// <param name="maxReps"></param>
+	  /// <param name="length"></param>
+	  /// <param name="constructorArgs"></param>
+	  /// <param name="description"></param>
+	  protected internal virtual void add(Type c, bool required, int maxReps, int length, Object[] constructorArgs,
+		  string description)
+	  {
+		 if (!typeof(IType).IsAssignableFrom(c))
+		 {
+			throw new HL7Exception("Class " + c.FullName + " does not inherit from " + "ca.on.uhn.datatype.Type",
+				HL7Exception.APPLICATION_INTERNAL_ERROR);
+		 }
+
+		 _items.Add(new AbstractSegmentItem(c, required, maxReps, length, constructorArgs, description));
+	  }
+
+	  /// <summary>
+	  /// Remove a valid index from a repeatable field.
+	  /// </summary>
+	  /// <param name="fieldNum">Repeatable field number</param>
+	  /// <param name="index">0-based index to be removed</param>
+	  public void RemoveRepetition(int fieldNum, int index)

```

Comment:
@nkm8 I don't suppose u kind me asking for some unit tests please?

I have an outstanding issue to create some Contribution Guidence, and unit tests for new behaviour will be a recommendation.

Reply:
@milkshakeuk This code is over 3 years old and I don't work with Hapi/HL7 anymore. Plus, both functions boil down to `List.Remove` and `List.RemoveAt` - I don't see any benefit of adding unit tests here. If `_items` was a public property, these functions wouldn't even be needed.

Most of the logic is to pretty-print a more readable exception - you could remove all that extra logic and simply let the exception for an index-out-of-range bubble up instead.

References:
- [Hapi code](https://sourceforge.net/p/hl7api/code/764/tree/trunk/hapi-mvn/hapi-base/src/main/java/ca/uhn/hl7v2/model/AbstractSegment.java#l561)
- [Unit tests in practice](https://github.com/dib0/NHapiTools/pull/5/commits/b56ad8426caaabcd414073853892d861b0c74c53#diff-851daece19aaed3745b0a63aef48f32b8ebef273524044e9a10607ca6739b6c3R332-R342)

Reply:
@nkm8 fair enough, I would like to simplify nhapi at some point since there is a lot of unnecessary abstractionsl like unsay these boil down to lists.

But small steps right.

Reply:
@milkshakeuk No worries - I understand completely. If no one else has shown an interest/need in this specific feature, you can honestly just toss this PR - no offense taken :shrug: 

I had to look a lot of this up again to remind myself why it was needed - I worked around it internally with custom builds of nHapi and nHapiTools, so it isn't blocking anything.